src/hints/state.rs

```diff
@@ -0,0 +1,372 @@
+    insert_value_from_var_name(vars::ids::INITIAL_ROOT, commitment_info.previous_root, vm, ids_data, ap_tracking)?;
+    insert_value_from_var_name(vars::ids::FINAL_ROOT, commitment_info.updated_root, vm, ids_data, ap_tracking)?;
+
+    let preimage = commitment_info.commitment_facts;
+    exec_scopes.insert_value(vars::scopes::PREIMAGE, preimage);
+
+    let merkle_height = get_integer_from_var_name(vars::ids::MERKLE_HEIGHT, vm, ids_data, ap_tracking)?.into_owned();
+    let tree_height: Felt252 = commitment_info.tree_height.into();
+    assert_tree_height_eq_merkle_height(tree_height, merkle_height)?;
+
+    Ok(())
+}
+
+pub const PREPARE_PREIMAGE_VALIDATION: &str = indoc! {r#"
+	ids.edge = segments.add()
+	ids.edge.length, ids.edge.path, ids.edge.bottom = preimage[ids.node]
+	ids.hash_ptr.result = ids.node - ids.edge.length
+	if __patricia_skip_validation_runner is not None:
+	    # Skip validation of the preimage dict to speed up the VM. When this flag is set,
+	    # mistakes in the preimage dict will be discovered only in the prover.
+	    __patricia_skip_validation_runner.verified_addresses.add(
+	        ids.hash_ptr + ids.HashBuiltin.result)"#
+};
+
+pub fn prepare_preimage_validation(
+    vm: &mut VirtualMachine,
+    exec_scopes: &mut ExecutionScopes,
+    ids_data: &HashMap<String, HintReference>,
+    ap_tracking: &ApTracking,
+    _constants: &HashMap<String, Felt252>,
+) -> Result<(), HintError> {
+    let new_segment_base = vm.add_memory_segment();
+    insert_value_from_var_name(vars::ids::EDGE, new_segment_base, vm, ids_data, ap_tracking)?;
+
+    let preimage: HashMap<Felt252, Vec<Felt252>> = exec_scopes.get(vars::scopes::PREIMAGE)?;
+    let node = get_integer_from_var_name(vars::ids::NODE, vm, ids_data, ap_tracking)?.into_owned();
+    let node_values = preimage
+        .get(&node)
+        .ok_or(HintError::CustomHint("preimage does not contain expected edge".to_string().into_boxed_str()))?;
+
+    if node_values.len() != 3 {
+        return Err(HintError::CustomHint(
+            "preimage value does not appear to be a NodeEdge".to_string().into_boxed_str(),
+        ));
+    }
+
+    let length_addr = new_segment_base;
+    let path_addr: Relocatable = (new_segment_base + 1)?;
+    let bottom_addr: Relocatable = (new_segment_base + 2)?;
+
+    vm.insert_value(length_addr, node_values[0])?;
+    vm.insert_value(path_addr, node_values[1])?;
+    vm.insert_value(bottom_addr, node_values[2])?;
+
+    // TODO: prevent underflow (original hint doesn't appear to care)?
+    // compute `ids.hash_ptr.result = ids.node - ids.edge.length`
+    let res = node - node_values[0];
+
+    // ids.hash_ptr refers to SpongeHashBuiltin (see cairo-lang's sponge_as_hash.cairo)
+    // it is a struct with 6 felts and `result` is the 4th
+    let hash_ptr = get_relocatable_from_var_name(vars::ids::HASH_PTR, vm, ids_data, ap_tracking)?;
+    let hash_ptr_result: Relocatable = (hash_ptr + 3)?;
```

Comment:
Seems fragile.

Reply:
This should be fixed now, I merged the offset PR.